🚀 2단계 - 리팩터링(메뉴)

mj950425 opened the pull request:
안녕하세요 리뷰어님~

많이 부족하긴한데, 그래도 리뷰 요청드려봅니다.

잘 부탁드립니다!

mj19950425 committed:
feature : step1

mj19950425 committed:
feature : 리뷰 반영

- 상품의 가격 변경시에 이벤트 핸들링을 통한 조건에 따른 메뉴 비노출 기능 추가
- jpa repo Optional 제거
- 트랜잭션 어노테이션 추가
- vo data class를 class로 변경

mj19950425 committed:
feature : 메뉴 구현

mj19950425 committed:
feature : 메뉴 구현

mj19950425 committed:
feature : 메뉴 구현

// TODO

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt:
```diff
@@ -1,16 +1,87 @@
 ) {
+    fun createMenu(request: CreateMenuRequest): CreateMenuResponse {
+        val menuGroup =
+            menuGroupRepository.findMenuGroupById(request.menuGroupId)
+                ?: throw IllegalArgumentException("존재하지 않는 메뉴 그룹입니다.")
+        val menuProducts =
+            request.menuProducts.map { menuProduct ->
+                val product =
+                    productRepository.findProductById(menuProduct.productId)
+                        ?: throw IllegalArgumentException("존재하지 않는 상품입니다.")
```
중요한 부분은 아닌데 코멘트 남깁니다.

사실 중복이 발생하는 상황은 아니라 메서드 추출이 불필요하긴 한데 이런 경우 getMenuGroupById와 같은 프라이빗 메서드가 있으면 메서드의 가독성 향상에 도움이 될 것 같아요.

mj950425 replied:
좋은 코멘트이네요. 감사합니다. 반영해두겠습니다.

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt:
```diff
@@ -1,16 +1,87 @@
+        val menuProducts =
+            request.menuProducts.map { menuProduct ->
+                val product =
+                    productRepository.findProductById(menuProduct.productId)
```
메뉴의 상품 목록을 하나씩 조회하면서 DB 쿼리를 날리고 있네요. 메뉴 상품이 많지 않거나 요청이 적은 경우라면 성능 문제는 없겠지만 반복적인 데이터베이스 쿼리로 인해 부하가 높아질 것 같습니다.

findAllByProductIds와 같은 형태로 개선해보면 어떨까요?

mj950425 replied:
수정해두겠습니다!

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt:
```diff
@@ -1,16 +1,87 @@
 class MenuServiceV2(
     private val menuDisplayChecker: MenuDisplayChecker,
+    private val menuGroupRepository: MenuGroupRepositoryV2,
+    private val productRepository: ProductRepositoryV2,
```
productRepository를 직접 의존하고 있네요. 메뉴쪽 도메인에 상품 조회를 위한 인터페이스를 만들고 해당 구현체를 infra 패키지로 옮기는 형태로 메뉴와 상품 도메인간의 결합을 끊어볼 수 있을까요?

DDD에서 부패 방지 계층([Anti-Corruption Layer Pattern](https://junhyunny.github.io/architecture/pattern/anti-corruption-layer-pattern/)과 같은 키워드로 한번 학습해보셔도 좋을 것 같아요.

mj950425 replied:
수정해두겠습니다~

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt:
```diff
@@ -1,16 +1,87 @@
+        return menu.id
+    }
+
+    fun unDisplay(menuId: UUID): UUID {
```
중요한 부분은 아니지만, 다소 어색해보이는데 hideMenu 등의 메서드 이름을 짓는건 어떨까요?

mj950425 replied:
용어사전과 함께 수정해뒀습니다!

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt:
```diff
@@ -1,16 +1,87 @@
+        return menu.id
+    }
+
+    fun findAll(menuId: UUID) {
+        val menu = menuRepository.findAll()
+    }
```
엔티티가 그대로 외부로 반환되고 있습니다. 영속성 컨텍스트 등 여러 문제가 발생할 여지가 많아보이는데요. 변환 후 반환하는건 어떨까요?

mj950425 replied:
구현이 안되어있는 부분이 있었군요. 수정해두겠습니다.

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/dto/response/ChangeMenuPriceResponse.kt:
```diff
@@ -0,0 +1,19 @@
+    val price: BigDecimal,
+) {
+    companion object {
+        fun of(menu: MenuV2): ChangeMenuPriceResponse {
```
중요한 부분은 아니지만 이펙티브 자바와 같은 도서에서는 정적 팩터리 네이밍 가이드 시 하나의 파라미터인 경우 from, 여러개인 경우 of로 가이드하고 있어요.

대부분 개발자가 이렇게 알고 있을 것이라 해당 규칙에 맞춰서 작성하면 어떨까 생각되네요.

mj950425 replied:
좋은 코멘트 감사합니다. 반영해두겠습니다

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt:
```diff
@@ -1,16 +1,87 @@
+    }
+
     fun handleProductPriceChanged(event: ProductPriceChangedEvent) {
-        menuDisplayChecker.unDisplayIfNeed(event.productId, event.price)
+        menuDisplayChecker.handleProductPriceChanged(event.productId, event.price)
     }
```
이벤트 리스닝을 하는 부분도 응용 서비스에 속하긴 하지만 상품 변경 이벤트를 처리하는 부분을 명확히 드러낼 수 있으므로 분리하는건 어떨까요?

mj950425 replied:
좋네요~ 수정해두겠습니다.

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/infra/JpaRepositoryUtils.kt:
```diff
@@ -0,0 +1,10 @@
+import kitchenpos.tobe.menu.domain.repository.MenuRepositoryV2
+import java.util.UUID
+
+fun MenuRepositoryV2.findMenuByIdOrThrow(menuId: UUID): MenuV2 {
```
일반적으로 find는 nullable, get은 non null 타입을 반환하는 의미로 많이 사용됩니다.

```suggestion
fun MenuRepositoryV2.getMenuById(menuId: UUID): MenuV2 {
```

non null이므로 예외를 던진다는 의미도 내포되어 있어요. 위와 같이 네이밍하는건 어떨까요?

mj950425 replied:
수정해두겠습니다~ 좋은 리뷰 감사합니다!!

kyucumber commented on src/main/kotlin/kitchenpos/tobe/product/exception/menu/MenuNameException.kt:
```diff
@@ -0,0 +1,5 @@
+package kitchenpos.tobe.product.exception.menu
+
+class MenuNameException(
+    override val message: String?,
+) : IllegalArgumentException(message)
```
예외를 만들어주셔서 코멘트 남겨요.

도메인에서 적절한 커스텀 예외를 만드는 것은 좋지만, 이 예외를 구분해 응답값을 다르게 한다거나 하는 니즈가 없다면 굳이 커스텀 예외를 만들 필요가 없어보여요.

현재 그런 부분은 보이지 않아 표준 예외인 IllegalArgumentException, IllegalState 등을 그대로 사용해도 되지 않을까 생각되네요. 이펙티브 자바나 코틀린 책에서도 표준 예외의 사용에 대한 내용을 다루고 있으니 한번 찾아보셔도 좋을 것 같네요. 물론 아래와 같은 경우는 커스텀 예외 사용을 권장합니다.

- 컨트롤러 레이어에서 JPA의 Dataintegrityviolationexception 과 같은 저수준 예외를 활용하는 경우
- IllegalArgumentException 등의 표준 예외가 이미 구현된 예외 처리 구문과 겹쳐 예외를 구분해야 하는 경우

mj950425 replied:
실무에서도 의미없는 예외를 만드는 경우가 종종 있었는데, 좋은 리뷰 감사합니다. 한번 리서치해보고 앞으로는 의식적으로 필요할때만 만들어야겠네요.

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드려요.

mj19950425 committed:
refact : 리뷰 반영

- find, get 맥락 분리
- findById -> findAll로 디비 성능 개선
- acl 구현
- undisplay -> hide
- findAll 미구현 부분 수정
- 정적팩토리 네이밍 수정, of -> from
- 응용 계층의 service와 handlerService 분리
- 의미없는 예외 클래스 제거

mj19950425 committed:
refact : 메소드이름, 클래스이름 변경

mj19950425 committed:
리팩토링 : 상품 아이디에 해당하는 상품을 전부 가져온다는 의미에서 fetch->get으로 메소드 명 변경 및 예외 로직 이동

kyucumber commented on src/main/kotlin/kitchenpos/tobe/menu/infra/ProductClientImpl.kt:
```diff
@@ -10,12 +10,13 @@ import java.util.*
+    override fun getProductPrices(productIds: List<UUID>): List<ProductPrice> {
+        val products = productRepository.findAllByIdIn(productIds)
+
+        if (products.size != productIds.size) {
+            throw IllegalArgumentException("상품이 존재하지 않습니다.")
```
중요한 부분은 아닐 수 있으나, 어떤 상품이 누락되었는지 예외 메시지에 식별자 정보가 남아야 문제 파악이 가능할 것 같네요.

kyucumber reviewed (approved):
간단한 코멘트 남겨두었으니 확인 부탁드려요.

다음 단계 진행을 위해 머지하겠습니다.
